### ISSUE TITLE
Unable to build PurchasesHybridCommon

### ISSUE THREAD

**Issue by CouplenessTech** [RevenueCat/react-native-purchases#200]
**Describe the bug**

Production build of the app fails with the following error:

> could not build Objective-C module 'PurchasesHybridCommon'

with the underlying error:

> /ios/Pods/PurchasesHybridCommon/ios/PurchasesHybridCommon/PurchasesHybridCommon/RCPurchases+HybridAdditions.h:12:12: error: declaration of 'RCPurchases' must be imported from module 'Purchases' before it is required

** Context **
Platform: Azure DevOps Hosted Pool (macos-10.15)
SDK: react-native-purchases@4.0.0 (latest)
Software: XCode 12.2

** Details **
Building using the command line:
/xcodebuild -sdk iphoneos -configuration Release -workspace /ios/XX.xcworkspace -scheme XX build CODE_SIGN_STYLE=Manual CODE_SIGN_IDENTITY=iPhone Distribution: XX (XX) PROVISIONING_PROFILE=XX PROVISIONING_PROFILE_SPECIFIER=

**Comment by aboedo** [comment]
hi! thanks for reporting. 

The problem seems to be when installing the pod. I've seen some weird behavior with cocoapods when there's some leftover cached data...
Could you try: 

- cleaning the derived data folder (https://programmingwithswift.com/delete-derived-data-xcode/)
- running `pod cache clean --all`
- `pod install`

And seeing if that fixes it, or the error message changes?

**Comment by CouplenessTech** [comment]
Hi, thanks for quick response.

Yes the build server is currently cleaning the cache. Although the error message says "declaration of 'RCPurchases' must be imported from module 'Purchases' before it is required" which indicates an actual error in the SDK. Still hoping cleaning the pod cache will resolve the issue though.

**Comment by aboedo** [comment]
I've seen that error before and it could be related to a few things, sometimes caching, sometimes having `Enable Modules` set to `NO`, or messing with the framework search paths, which would all make the module `Purchases` not available for lookup when compiling. 
If cleaning up the cache does not solve it, it might be related to something else, but very likely a linking issue. 
Out of curiosity, does this also happen when building locally? Are you building `Release` or `Debug`?

**Comment by CouplenessTech** [comment]
Cleaning the cache did not help on the build server. It builds locally with `Debug` build and XCode 12.3 but not `Release` and Xcode 12.2. The build server was able to build the app before upgrading the RevenueCat SDK, which was the only change, upgrade from 3.4.2 to 4.0.0.

Not familiar with `Enable Modules` but we are using other modules than react-native-purchases.

**Comment by tarouboy** [comment]
+1
couldn't build / archive after upgrading to Xcode Version 12.3

**Comment by CouplenessTech** [comment]
Still unable to build, which hinders us from enabling the new purchase flow SCA (new EU legislation), major issue.

Have tried multiple things, including:

- `pod cache clean --all`
- `xcodebuild clean`
- removing Pods directory
- removing CocoPods cache directory
- removing Podfile.lock
- removing DerivedData directory
- removing build directory

Still the same error.

@tarouboy what did you do to make it work?

@aboedo any thoughts?

**Comment by aboedo** [comment]
@CouplenessTech I haven't been able to reproduce yet, I'm not sure exactly how to do it. 
A couple of clarifications in the meantime:
This shouldn't stop you from being compatible with SCA:
- Plugin versions >= `3.4.0` all handle pending / deferred purchases the same way
- `4.0.0` only adds a way to access `simulatesAsksToBuyInSandbox` to make testing slightly easier, but as long as you're handling errors with `pending` or `cancelled` states in your purchase flow, the app should work correctly. Even with `simulatesAsksToBuyInSandbox`, testing this is less-than-ideal, since Apple doesn't provide a way for the purchase to eventually get approved, so pending purchases are just left hanging when testing (they never get confirmed).
- SCA is enabled by Apple, not apps, so it's already in use. You need an European account to be able to test it out. 
- More information about SCA here: https://www.revenuecat.com/blog/strong-customer-authentication

As for root causes for this: 
- Version 3.4.3 used the pod statically, but this was later reverted because it caused some build issues for other users. 
- Version 4.0.0 moved our dependency to ReactNative from `React` to `React-Core`, which has some different restrictions for linking. 

Since it was working for you on 3.4.2, maybe re-enabling static linking for the pod would help. 

i.e.:

```ruby
pod 'RNPurchases', :path => '../node_modules/react-native-purchases', use_frameworks! :linkage => :static
```
Let me know if that helps!

**Comment by CouplenessTech** [comment]
RevenueCat SDK `3.4.2` only returns `Unknown Error` when SCA is initiated, therefore we cannot know if there actually was an unknown error or that the user have initiated SCA. `Purchases.purchasePackage(...)` does not give any information of `pending` or `cancelled`, only throws the `Unknown Error`.

Our users are unable to complete SCA. They accept everything during the bank webpage authentication but no subscription is registered within RevenueCat, looking at the RevenueCat dashboard, and no app subscription is registered on the users Apple ID / Google account.

We do not use `use_frameworks!` and haven't had the need for declaring any pods in the Podfile, except a couple from react-native-permissions, the rest are auto-linked. Here is our Podfile:

```
$deployment_target = '10.0'
platform :ios, $deployment_target
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

target 'Coupleness' do
  config = use_native_modules!
  use_react_native!(:path => config["reactNativePath"])

  permissions_path = '../node_modules/react-native-permissions/ios'
  pod 'Permission-Contacts', :path => "#{permissions_path}/Contacts.podspec"
  pod 'Permission-PhotoLibrary', :path => "#{permissions_path}/PhotoLibrary.podspec"

  def set_deployment_target_for_all_pods(installer)
    installer.pods_project.targets.each do |t|
      t.build_configurations.each do |config|
        config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = $deployment_target
      end
    end
  end

  target 'CouplenessTests' do
    inherit! :complete
    # Pods for testing
  end

  use_native_modules!

  # Enables Flipper and forces correct version of pods
  use_flipper!
  post_install do |installer|
    flipper_post_install(installer)
    set_deployment_target_for_all_pods(installer)
  end
end
```

As a part of React Natives own upgrade guide, we removed `use_frameworks!` long ago. Currently we are using React Native `0.63.3`. Do you still suggest that we add both `use_frameworks!` and `pod 'RNPurchases', :path => '../node_modules/react-native-purchases'`?

**Comment by aboedo** [comment]
I'm a bit confused about the podfile... Do you have the plugin installed through cocoapods or through the manual install path? I'm asking because I don't see an entry for `RNPurchases`, but I assumed you had `RNPurchases` installed through cocoapods, which may have been a wrong assumption

**Comment by CouplenessTech** [comment]
We followed your installation guide a few months ago. `use_native_modules!` enables auto-linking, thus we do not need to specify every pod in the podfile, React Native takes care of it. We are using several other Pods, including `AppsFlyerFramework` and `RNPermissions` etc. Every plugin is installed using standard `npm i [plugin]` and rare cases need some extra config, e.g. `react-native-permissions` needed the three lines shown in the podfile above.

**Comment by CouplenessTech** [comment]
Using Cocoapods: 1.10.0 (from podfile.lock)

**Comment by CouplenessTech** [comment]
I tried adding `pod 'RNPurchases', :path => '../node_modules/react-native-purchases'`, no change in error message.

I tried adding both `use_frameworks! :linkage => :static` and `pod 'RNPurchases', :path => '../node_modules/react-native-purchases'`, resulted in multiple new errors, here comes some of them:

> 2021-01-12T17:45:10.9968880Z Build system information
2021-01-12T17:45:10.9986540Z warning: Multiple targets match implicit dependency for linker flags '-framework DoubleConversion'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'Coupleness' from project 'Coupleness')
2021-01-12T17:45:10.9988300Z 
2021-01-12T17:45:10.9989000Z Build system information
2021-01-12T17:45:10.9990580Z warning: Multiple targets match implicit dependency for linker flags '-framework folly'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'Coupleness' from project 'Coupleness')
2021-01-12T17:45:10.9991620Z 
2021-01-12T17:45:10.9992220Z Build system information
2021-01-12T17:45:10.9993790Z warning: Multiple targets match implicit dependency for linker flags '-framework glog'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'Coupleness' from project 'Coupleness')
2021-01-12T17:45:10.9994790Z 
2021-01-12T17:45:11.0235940Z error: ReactCommon does not support provisioning profiles. ReactCommon does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'ReactCommon' from project 'Pods')
2021-01-12T17:45:11.0240330Z error: PurchasesCoreSwift does not support provisioning profiles. PurchasesCoreSwift does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'PurchasesCoreSwift' from project 'Pods')
2021-01-12T17:45:11.0250350Z error: libwebp does not support provisioning profiles. libwebp does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'libwebp' from project 'Pods')
2021-01-12T17:45:11.0256200Z error: ReactNativeWebPFormat does not support provisioning profiles. ReactNativeWebPFormat does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'ReactNativeWebPFormat' from project 'Pods')
2021-01-12T17:45:11.0266380Z error: React-RCTImage does not support provisioning profiles. React-RCTImage does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'React-RCTImage' from project 'Pods')
2021-01-12T17:45:11.0272770Z error: RNInAppReviewIOS does not support provisioning profiles. RNInAppReviewIOS does not support provisioning profiles, but provisioning profile beta has been manually specified. Set the provisioning profile value to "Automatic" in the build settings editor. (in target 'RNInAppReviewIOS' from project 'Pods')

**Comment by CouplenessTech** [comment]
@aboedo What did you find confusing about the podfile? something you thought were missing? something new/unknown?

**Comment by aboedo** [comment]
@CouplenessTech thanks for the answers, I was just confused because I wasn't sure whether you were using cocoapods or manual install and forgot to ask. 

yeah, it looks like some of the other frameworks have some issues when doing static linking. This seems kinda similar to #198, maybe the solutions there help (they involve setting `use_modular_headers!`, but disabling it for problematic pods via adding `modular_headers: false` to them). 

I'm sorry about the installation issues, there are a lot of configurations from different developers and it's often tricky to find which solutions work for each app. The `4.0.0` changes were done because the `React` dependency didn't compile correctly under some configurations on Xcode 12, and it led us to realize that we actually need to be using `React-Core`, which comes with its own restrictions when it comes to compilation.

**Comment by CouplenessTech** [comment]
@aboedo Well I'm glad you're trying to help out.

I've queued up three new builds, one with `use_modular_headers!, one with `pod 'React-Core', :path => '../node_modules/react-native/', :modular_headers => true` and a third with `pod 'RNPurchases', :path => '../node_modules/react-native-purchases', :inhibit_warnings => true, :modular_headers => false`, all taken from the issue you linked. Hopefully one of them will work.

**Comment by aboedo** [comment]
@CouplenessTech awesome! keep me posted!

**Comment by CouplenessTech** [comment]
@aboedo I will

**Comment by CouplenessTech** [comment]
@aboedo All three failed.

`use_modular_headers!` generated a new error:
> /ios/Pods/Headers/Public/glog/glog/logging.h:512:1: fatal error: import of module 'glog.glog.log_severity' appears within namespace 'google'

But no improvement, as far as I know.

The other two builds generated the old error.
> declaration of 'RCPurchases' must be imported from module 'Purchases' before it is required

**Comment by aboedo** [comment]
@CouplenessTech for the error with `use_modular_headers!`, you can disable the setting for problematic pods. 
In my testing, I had issues with `glog`, `DoubleConversion` and `folly`, which I solved by doing:


```ruby
pod 'DoubleConversion', :podspec => '../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec', :modular_headers => false
pod 'glog', :podspec => '../node_modules/react-native/third-party-podspecs/glog.podspec', :modular_headers => false
pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec', :modular_headers => false
```

**Comment by CouplenessTech** [comment]
@aboedo okey I'll try that, those three are the only ones we have in `/node_modules/react-native/third-party-podspecs`

**Comment by CouplenessTech** [comment]
@aboedo back to the original error

**Comment by CouplenessTech** [comment]
@aboedo Maybe this warning tell you something?

> /ios/Pods/PurchasesHybridCommon/ios/PurchasesHybridCommon/PurchasesHybridCommon/RCPurchaserInfo+HybridAdditions.h:6:2: warning: missing submodule 'Purchases.Purchases'

**Comment by aboedo** [comment]
@CouplenessTech thanks for the update. the warning matches the error messages: it looks like what's happening is that for some reason, the `PurchasesHybridCommon` can't find the `Purchases` module, which it depends on. 

I'm still not entirely sure what's going on, but I'm going to try a few things to see if I can reproduce and figure this out. I think this might either be an issue with the framework search paths, with the creation of modules / module headers, or something related.

**Comment by aboedo** [comment]
@CouplenessTech update: 

I just tried installing the sdk again with a brand new app, and it worked, generating the following Podfile: 
```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '10.0'

target 'AwesomeProject' do
  config = use_native_modules!

  use_react_native!(:path => config["reactNativePath"])

  pod 'RNPurchases', :path => '../node_modules/react-native-purchases'

  target 'AwesomeProjectTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable these next few lines.
  use_flipper!
  post_install do |installer|
    flipper_post_install(installer)
  end
end

target 'AwesomeProject-tvOS' do
  # Pods for AwesomeProject-tvOS

  target 'AwesomeProject-tvOSTests' do
    inherit! :search_paths
    # Pods for testing
  end
end

```

I also tried generating a new plugin, to make sure that we were setting dependencies correctly, and they match the ones in our plugin. 

After that, I tried yet another project, using `npm i react-native-purchases` (without `--save`), and I was able to reproduce your error 🎉 

So it seems like the current version of the plugin isn't compatible with auto-linking. 

I'm going to work on fixing that, but in the meantime, could you try removing the plugin and re-installing via: 
```bash
npm uninstall react-native-purchases
npm install --save react-native-purchases
npx react-native link react-native-purchases
```
?

This fixed the issues on my test project.

**Comment by CouplenessTech** [comment]
@aboedo Great that you managed to recreate the issue. I already tried to reinstall the package, and link it, but I'll try your set of commands anyways

**Comment by CouplenessTech** [comment]
Tried yours and a few different versions of your commands, all failed, same old error

**Comment by aboedo** [comment]
I'm pretty confused by this... do the pods get installed correctly? 
Do they show up in the Pods project? Specifically `PurchasesCoreSwift`, `Purchases` and `PurchasesHybridCommon`
<img width="256" alt="image" src="https://user-images.githubusercontent.com/3922667/104633257-a7243200-567d-11eb-9be6-d4ded59ab6f9.png">

**Comment by CouplenessTech** [comment]
Yes the show up in the Pods project.

Have also run `pod update [name]` on each of the four `PurchasesCoreSwift`, `Purchases`, `PurchasesHybridCommon` and `RNPurchases` but didn't have any effect.

Running `npx react-native link react-native-purchases` said that iOS was already linked, even if I first ran:
1. `npm uninstall react-native-purchases`
2. Removed `pod 'RNPurchases', :path => '../node_modules/react-native-purchases'` from podfile
3. `pod update` (which outputs that purchases has been removed)
4. `npm i react-native-purchases`
5. `pod install`
6. `pod update Purchases`
7. `pod update PurchasesCoreSwift`
8. `pod update PurchasesHybridCommon`
9. `pod update RNPurchases`

**Comment by aboedo** [comment]
- is `Enable Modules` set to `YES` in the project? 
- any chance that framework search paths are being modified? 
- I think you mentioned doing this before, but in my case, I needed to explicitly save the pod to the Podfile by doing `npm i --save react-native-purchases`
- is the deployment target correctly set for the project, not just the podfile?
- any other warnings or output from xcode that might seem relevant?

**Comment by CouplenessTech** [comment]
1. Yes `Enable Modules` is set to `YES`
2. Not that I'm aware of, anything you had in mind that could do that?
3. The `--save` argument for `npm install` is deprecated since v5 and now no longer part of the official docs. But yes I have tried to add the pod to podfile.
4. Yes it is, was `10.0` but have also tried `12.0` as that was a solution for somebody else.
5. XCode outputs about 100.000 lines during build but no I have not found anything else that might be relevant

I tried degrading back to `react-native-purchases@3.4.2` and it builds.

**Comment by aboedo** [comment]
does 3.4.3 also build? there are some changes in that one as well that might be related. 

for 2., I've seen some build systems that override settings. For example Flutter adds some extra steps that modify the search paths, so for that one we had to explicitly re-add the frameworks so the search paths.

**Comment by CouplenessTech** [comment]
I have not tried `3.4.3` but I will.

Framework Search Paths are not configured, was not necessary for `3.4.2` and auto-linking. I'll try adding `$(PROJECT_DIR)/../node_modules/react-native-purchases/ios` to all configurations, taken from your manual installation guide.

**Comment by CouplenessTech** [comment]
Upgraded to `4.0.0` and added `Framework Search Paths`, same error as before.
`3.4.3` is queued on the build server...

**Comment by CouplenessTech** [comment]
`3.4.3` did not build, same error as `4.0.0`

**Comment by aboedo** [comment]
@CouplenessTech 
Interesting, thanks for the update! 

I think that might actually help narrow down the problem. 

The main difference between `3.4.2` and `3.4.3` in terms of compilation and linking is that `3.4.2` forced the usage of static frameworks, whereas `3.4.3` uses dynamic by default. In static frameworks, linking happens statically in an earlier stage, so there might be an issue either with linker flags that the framework needs but aren't set (like `-ObjC`, `-all_load`), with the framework search paths not being set correctly, or something related. 

It'd be really hard for me to figure this out without actually testing the project, since these settings are very specific, however, you can force `3.4.2` behavior in other versions of the SDK by setting in your Podfile:

```ruby
use_frameworks! :linkage => :static
```

This might force you to set `use_modular_headers!` as well for `React-Core` to work. 
If this causes the `modular_headers` error that was mentioned earlier in this thread, then that can be fixed by disabling `modular_headers` individually on the problem pods, like `DoubleConversion, `glog`, `Folly`. 

Hope this helps!

**Comment by CouplenessTech** [comment]
Tried them separately before and now together, same error

```
  use_frameworks! :linkage => :static
  use_modular_headers!
  pod 'DoubleConversion', :podspec => '../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec', :modular_headers => false
  pod 'glog', :podspec => '../node_modules/react-native/third-party-podspecs/glog.podspec', :modular_headers => false
  pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec', :modular_headers => false
```

Because of the build script specifying provisioning profile, I also needed to add the following:
```
  # Stop Cocoapods from signing frameworks
  # REF: https://stackoverflow.com/a/55750843/14454546
  def disable_provisioning_profile_for_pods(installer)
    installer.pods_project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings['EXPANDED_CODE_SIGN_IDENTITY'] = ""
        config.build_settings['CODE_SIGNING_REQUIRED'] = "NO"
        config.build_settings['CODE_SIGNING_ALLOWED'] = "NO"
      end
    end
  end
```
and call it in `post_install` but still does not work, got errors with Flipper (use_frameworks and Flipper are not compatible), I'll try disabling Flipper and see how that goes, but I should not need to do that.

**Comment by aboedo** [comment]
getting the same error with those changes is really unexpected, I'd at least have expected a new error. 
any chance that something is getting cached in a way that prevents the changes from being correctly reflected?

**Comment by CouplenessTech** [comment]
The issue remains. Cleared several caches multiple times during the troubleshooting.
@aboedo have you made any progress?

**Comment by aboedo** [comment]
@CouplenessTech I have still not been able to reproduce the issue (other than when I tried not doing `--save`). 
I'm hitting a bit of a wall when trying to find other ways to reproduce it. 

could you provide some repro steps? does this happen on a new project for you? do the issues only happen on CI? 
could you provide some more info, regarding: 
- rn version
- podfile, cocoapods version

**Comment by CouplenessTech** [comment]
@aboedo Sorry for the delay. I have already provided you with rn version, podfile and cocoapods version, scroll up through this issue. The build server is now testing a new project and also the old project where RevenueCat got stripped all together and then reinstalled according to your docs (with --save even though it is deprecated). Time will tell if any of those builds succeeds, I'll get back to you on that.

**Comment by iDuuck** [comment]
Unfortunately I have the same error. Overtook the project from another dev, but the previous developer didn't had any problems before. Also → Only happening in Release. Perhaps some sort of Build Settings?

**Comment by EricWiener** [comment]
Also having this issue when trying to archive a release build. Able to build no problem. Just the archiving is an issue.

**Comment by aboedo** [comment]
what error message are you seeing when archiving?

**Comment by EricWiener** [comment]
@aboedo the exact same one described by @CouplenessTech in the original post

**Comment by CouplenessTech** [comment]
Many tries later, still same error, even for new react native projects

**Comment by aboedo** [comment]
hey everyone, I'm still working on this, I'm having a hard time reproducing locally since it archives correctly for me on new projects. I can archive and even validate for App Store correctly. I'll keep trying things to reproduce, but if you have solid steps that reproduce this for you, please let me know because it'd help a ton in getting a fix in.

**Comment by aboedo** [comment]
it looks like react-native updated `React-Core` for Xcode 12 compatibility in react-native release `0.63.4`, back in November. https://github.com/facebook/react-native/commit/6e08f84719c47985e80123c72686d7a1c89b72ed

Could you try upgrading to see if it solves it? If it does, it'd explain a lot, since we switched to the `React-Core` dependency rather recently. @CouplenessTech @iDuuck @EricWiener

**Comment by aboedo** [comment]
This would be with `react-native-purchases` >= `4.0.0`

**Comment by aboedo** [comment]
you may have to remove flipper for it to work, I had to comment from my podfile since there are some issues with its Podspec

**Comment by CouplenessTech** [comment]
@aboedo Good that you are still working on this. We use React Native `0.63.3` and our Azure build server uses XCode 12.3, which was the latest at the time of configuration, but now they have made 12.4 available. I shall try a build with XCode 12.4.

**Comment by aboedo** [comment]
@CouplenessTech lmk how it goes! FWIW I think the important bit is upgrading to React Native `0.63.4`, since that sets the `ENABLE_MODULES` flag to `YES`, and not having had that might easily have been the problem, and I think RN might actually overwrite whatever value your project is set to

**Comment by valdrox** [comment]
I had this problem too and tried a few things and now the problem is gone. I reduced react-native-purchases to 4.0.0 from 4.0.1 and npx react-native unlink react-native-purchases, and after reinstalling dependencies the problem was gone. Hopefully one of these two things solves your issue too.

**Comment by CouplenessTech** [comment]
Neither upgrading to React Native 0.63.4 nor unlinking and reinstalling dependencies worked

**Comment by NiklasMencke** [comment]
I had the exact same problem.

React Native 0.63.4
react-native-purchases 4.0.1

Before i was using 3.xx and it was working without issues.
I then deleted the Pods folder, deleted the Podfile.lock and built again and it works now

**Comment by aboedo** [comment]
@NiklasMencke that makes sense, thanks for posting. 
@iDuuck @EricWiener were you able to get it working?

**Comment by EricWiener** [comment]
@aboedo I will try to check soon. Sorry for the delay

**Comment by EricWiener** [comment]
@aboedo the solution @NiklasMencke mentioned worked for me. Thanks!

**Comment by aboedo** [comment]
that's great to hear! thanks for reporting back!

**Comment by EricWiener** [comment]
@aboedo I spoke too soon. I was able to build, but when I just archived to try to release, it fails with the same error as before.

**Comment by louiechristie** [comment]
# Trying to archive in XCode fails for me too.

For anyone reading this to save you some time, here's all the things I tried. (And to make my last few hours seem worthwhile.)

I was working with an app I generated using Expo about a year ago and then ejected to be an Expo bare app a couple of months ago.

## Upgrade npm dependencies

```console
npx npm-check-updates -u
```
Reference: https://stackoverflow.com/a/16074029/3469524

Use compatible dependencies for Expo Bare App and Expo Web
```console
expo upgrade
```
(This will also remove and reinstall your node_modules folder and package-lock.json file for you.)
Reference: https://dev.to/expo/expo-sdk-40-is-now-available-1in0#bare-workflow

## Upgrade Cocoapods
```
sudo gem install Cocoapods

```
References: https://guides.cocoapods.org/using/getting-started.html
https://stackoverflow.com/questions/39481636/updating-to-latest-version-of-cocoapods/51103633#51103633

## Update Flipper

ios/Podfile
```ruby
use_flipper!('Flipper' => '0.78.0')
```

android/gradle.properties
```
FLIPPER_VERSION=0.78.0
```

Reference: https://github.com/facebook/react-native/issues/30836#issuecomment-789368146

## Update RNPurchases dependencies

```console
react-native link react-native-purchases
```

Manually update RNPurchases

```ruby
  pod 'RNPurchases', :path => '../node_modules/react-native-purchases', :inhibit_warnings => true, :modular_headers => false
```

Reference:
https://github.com/RevenueCat/react-native-purchases/commit/ef2b786dec9421c84045f5d038cc84f85434fbbe#r47791138

Look at your git diff and manually remove duplicates lines that have been generated in ios/Podfile and android/app/src/main/java/[your/package/name]/MainApplication.java

## Reinstall iOS Dependencies
```console
rm ios/Podfile.lock
pod install --repo-update
```

## Clean React Native build files
```
cd android && ./gradlew clean && cd ../
cd ios && xcodebuild clean && cd ../
```
Reference: https://stackoverflow.com/a/56941424/3469524

## Remove old version of the app from your iPhone
On home screen click and hold icon -> 'Remove App' -> 'Delete App'

## Re-pair your device in Xcode

In Xcode go to Window -> Devices and Simulators right click your device click 'Unpair device'. Then unplug the device and plug it back in again and tap to trust it when popup displays.

## Run in Xcode
Open in Xcode at the top if it says Any iOS Device (arm64, armv7) click on it and select your actual iPhone device. (Note: you need an actual iPhone, you can't use a simulator with react-native-purchases.) Click the big play button ▶

## [optional] Ask yourself about your life choices when you next upgrade to a version x.0.0 or x.0.1 of a piece of software and wether you can wait a bit for someone else to be the guinea pig.

## [optional] While waiting for React Native to build watch this video by the inventor of WhatsApp's Erlang language on 'the mess we're in' with dependencies and fantasise about building his condenser idea:
https://www.youtube.com/watch?v=lKXe3HUG2l4

## Build in Release mode

Product -> Scheme -> Edit Scheme click on 'Run' on the left, change Build Configuration to 'Release'
Press that play button again ▶

## Get original error
> error: declaration of 'RCPurchases' must be imported from module 'Purchases' before it is required

## [optional] Despair. Really question those life choices

## Actually read through all the comments on this issue and found that someone had some luck with this command and run it:
```
pod cache clean --all
```
Reference: https://github.com/RevenueCat/react-native-purchases/issues/200#issuecomment-758538322

## Build in release mode
Press that play button again ▶
It works!

## Make archive in Xcode 

Product -> Archive

## Get same error as before.

**Comment by louiechristie** [comment]
# Update: it worked for me 🎉

## Was able to make an archive in Xcode by reinstalling Pods (iOS dependencies Cocoapods)
```
cd ios
rm -rf Pods
rm Podfile.lock
pod install 
```
Then I was then able to make an archive in XCode
XCode -> Product -> Archive

Reference: I did this after I read this comment by @NiklasMencke :
https://github.com/RevenueCat/react-native-purchases/issues/200#issuecomment-788065699

I don't know why this works. I thought previously deleting Podfile.lock would be enough. I speculate either:
* caching issue that is only resolved by deleting Pods folder
* race condition loading dependencies?

Also see: [all the things I did first before this worked for me](https://github.com/RevenueCat/react-native-purchases/issues/200#issuecomment-789793352)

**Comment by aboedo** [comment]
Closing out this issue due to inactivity. If you're still running into issues, try following [our SDK troubleshooting guide](https://docs.revenuecat.com/docs/troubleshooting-the-sdks) and open up a new issue if the steps described there don't help.

**Comment by therunbuddy** [comment]
I am still getting this issue on my repo. Every time I want to do a release build I have to delete the pods and re-install. Did anyone solve this issue permanently without having to do ```cd ios
rm -rf Pods
rm Podfile.lock
pod install ```

**Comment by imekinox** [comment]
This is something I've been experiencing for a while, it starts working after doing some random changes (listed above ) and spending days to fix it, then out of the blue, it starts failing again. This happens on our CI/CD pipeline without any big changes to Xcode config or code. 

No other library happens to behave like this one. Have any of you moved out to something else?

**Comment by imekinox** [comment]
So we could use the native SDKs? That's an option we are considering, we have a bunch of native integrations so we could pull this off too. Do you think it's a better approach?

**Comment by aboedo** [comment]
You can use the native SDKs directly, of course, although it will take some extra work. All of what `react-native-purchases` does is to bridge over the native SDKs to React-Native, so you'd essentially be doing the same as what `react-native-purchases` does

**Comment by kdthomas2121** [comment]
Is there an update to this @aboedo? I'm still seeing it almost every month

**Comment by aboedo** [comment]
@kdthomas2121 could you clarify exactly what you're running into? We've made quite a few changes since this issue was posted, including migrating the entire internal dependency that was originally causing the issue into Swift. 
So if you're still seeing it on modern SDK versions, it's likely to be caused by something different. 

Would you mind opening up a new issue explaining what you're seeing?

**Comment by LydGol90** [comment]
Hi there, I wanted to share here in case this helps anyone else or helps debug the issue further. 

I had never experienced this issue before with my project which has been using this lib for over 2 years. Then I got a new macbook with M2 Max chip and installed latest xcode 14.3.0 and got this for the first time. Others on here have said it happens randomly for them but I have found a specific scenario where I can replicate every time. 

With a fresh clean project, I run the pod install and everything works fine, I can run over and over, I can make changes to podfile, install other libraries and run pod install again etc and everything continues to work fine. The issues begin if I run the xcode clean command (command + shift + K). When I run this and try to build it will fail every time with the `could not build Objective-C module 'PurchasesHybridCommon'` error. Then the only way to recover is to delete pods and pod install again. (FYI for me it is enough to just delete the Purchases, PurchasesCoreSwift and PurchasesHybridCommon folders and run pod install, I do not have to delete and re-install all my Pods).

I'm not sure exactly what the xcode clean is doing under the hood but it is the only thing that causes this issue to happen for me. I can delete `ios/build` folder, I can delete everything under the xcode `DerivedData` folder and do a clean build with with no errors.  

Unfortunately I don't have time to investigate further due to other deadlines so I have simply been avoiding using the xcode clean command. Hope this helps others

**Comment by fellipetav** [comment]
The team responsible for Flutter projects here encountered this problem today. We still haven't figured out what is a definitive solution. But we noticed that it's happening only when running on iOS Simulators. Doesn't happen when running (debugging or not) in physical ones.

The problem also started after we cleaned the build folder.

Even after deleting Pods and Podfile.lock then running pod install, the problem persists. The only difference is that we can build and run the project, but the error comes up always when we face any kind of exception. But as I said before, it happens only in iOS Simulator.
